metrics/sinks/statsd/statsd_client.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package statsd
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/golang/glog"
+	"net"
+)
+
+type statsdClient interface {
+	open() error
+	close() error
+	send(messages []string) error
+}
+
+type statsdClientImpl struct {
+	host             string
+	numMetricsPerMsg int
+	conn             net.Conn
+}
+
+func (client *statsdClientImpl) open() error {
+	var err error
+	client.conn, err = net.Dial("udp", client.host)
```

Comment:
is only UDP supported?

Reply:
AFAIK, some implementations, i.e., [the etsy statsd](https://github.com/etsy/statsd/blob/master/docs/server.md), of statsd protocol supports tcp but not all of them because it is not part of the statsd protocol.

But according to the [introduction page of statsd](https://codeascraft.com/2011/02/15/measure-anything-measure-everything/):
> So, why do we use UDP to send data to StatsD? Well, it’s fast — you don’t want to slow your application down in order to track its performance — but also sending a UDP packet is fire-and-forget. Either StatsD gets the data, or it doesn’t. The application doesn’t care if StatsD is up, down, or on fire; it simply trusts that things will work. If they don’t, our stats go a bit wonky, but the site stays up.


udp is preferred and dup server is the default server for statsd.

Reply:
you are right @andyxning , tcp is not part of statsd protocol and udp is the default for statsd.